compiler/strings.go

```diff
@@ -0,0 +1,270 @@
+
+	var runes []rune
+	for _, r := range s {
+		if !isAlphaNum(r) {
+			if !wasUnderscore {
+				runes = append(runes, '_')
+				wasUnderscore = true
+			}
+			continue
+		}
+		wasUnderscore = false
+		runes = append(runes, r)
+	}
+
+	for len(runes) > 0 && runes[0] == '_' {
+		runes = runes[1:]
+	}
+	for len(runes) > 0 && runes[len(runes)-1] == '_' {
+		runes = runes[:len(runes)-1]
+	}
+
+	// pass 2: for each consecutive upper case characters, insert
+	// a break between the last upper case character and its
+	// predecessor
+	var wasUpper int
+	var buf bytes.Buffer
+	for i := 0; i < len(runes); i++ {
+		if !isAlphaNum(runes[i]) {
+			if !wasUnderscore {
+				buf.WriteRune('_')
+				wasUnderscore = true
+			}
+			continue
+		}
+
+		// if it's upper cased, check if we have a succession of
+		// uppercase letters.
+		if unicode.IsUpper(runes[i]) {
+			if wasUpper == 0 && buf.Len() > 0 && !wasUnderscore {
+				buf.WriteRune('_')
+			}
+			wasUpper++
+		} else {
+			if wasUpper > 1 && buf.Len() != 1 {
+				if len(runes) > 1 && runes[i-2] != '_' && runes[i-1] != '_' {
+					buf.Truncate(buf.Len() - 1)                // remove last upper case letter
+					buf.WriteRune('_')                         // insert rune
+					buf.WriteRune(unicode.ToLower(runes[i-1])) // re-insert last letter
+				}
+			}
+			wasUpper = 0
+		}
+		wasUnderscore = false
+		buf.WriteRune(unicode.ToLower(runes[i]))
+	}
+
+	return buf.String()
+}
+
+func camelCase(s string) string {
+	var first = true
+	var wasUnderscore bool
+	var buf bytes.Buffer
+	for _, r := range s {
+		// replace all non-alpha-numeric characters with an underscore
+		if !isAlphaNum(r) {
+			r = '_'
+		}
+
+		if r == '_' {
+			wasUnderscore = true
+			continue
+		}
+
+		if first || wasUnderscore {
+			r = unicode.ToUpper(r)
+		}
+		first = false
+		wasUnderscore = false
+		buf.WriteRune(r)
+	}
+
+	return buf.String()
+}
+
+// takes strings like "foo bar baz" and turns it into "foobarbaz"
+// if title is true, then "FooBarBaz"
+func concatSpaces(s string, title bool) string {
+	var buf bytes.Buffer
+	var wasSpace bool
+	for i, r := range s {
+		if unicode.IsSpace(r) {
+			wasSpace = true
+			continue
+		}
+		if i == 0 || (wasSpace && title) {
+			r = unicode.ToUpper(r)
+		}
+		buf.WriteRune(r)
+		wasSpace = false
+	}
+	return buf.String()
+}
+
+func packageName(s string) string {
+	return cleanCharacters(strings.ToLower(concatSpaces(s, false)))
+}
+
+func normalizeServiceName(s string) string {
+	return cleanCharacters(concatSpaces(s, true) + "Service")
+}
+
+func cleanCharacters(input string) string {
+	var buf bytes.Buffer
+	for _, r := range input {
+		// anything other than a-z, A-Z, 0-9 should be converted
+		// to an underscore
+		if !isAlphaNum(r) {
+			r = '_'
+		}
+		buf.WriteRune(r)
+	}
+	return buf.String()
+}
+
+func normalizeEndpointName(e *openapi.Endpoint) string {
+	if opID := e.OperationID; len(opID) > 0 {
+		return operationIDToName(opID)
+	}
+
+	path := strings.TrimSuffix(e.Path, ".json")
```

Comment:
Oh I get it... So in `fixtures/catsandogs.yaml` you have `/cats.{proto|json}` should this be normalized to `GetCats`, and not `GetCatsProtoJson` ?

Reply:
That's the hope! In our frameworks (marvin/gizmo), it's pretty easy to expose the same go-kit endpoint.Endpoint on multiple transports.

Here's another small example service that actually has `cats.{format}` as the path: https://github.com/jprobinson/cats-and-appengine-part-3/blob/master/service.yaml#L15

Instead of `GetCatsFormat`, it'd give us a nice `GetCats`.

Reply:
right on. I'll apply the necessary fixes